### ISSUE TITLE
Enable installation of "deb" Nexus 7 2013 gsm 

### ISSUE THREAD

**Issue by doniks** [ubports/ubuntu-touch#643]
### Description of the feature
Enable installation of "deb" Nexus 7 2013 gsm 

There is no porting needed. Not even compiling images. All the pieces are on the image server already. I think the only thing missing is some glue in the json files on the server or maybe in the installer. 

The "deb" is the sibling of the "flo" device. The flo is the wifi version, deb is gsm. 

### Steps to reproduce
- use ubports-installer
- try to install for deb

### Expected behavior
Can be installed

### Actual behavior
deb is not an option in the installer 

### Workaround 
NB: some steps are written from memory, syntax might be slightly different 
- install flo images with the installer
- boot, look around, 
=> observe that basically everything works (except sound)
- manually download the deb device image from image server
  - http://system-image.ubports.com/pool/device-1b9d2f7256c3d6c98ba93b66ad48d9ce75833978cd65364df45693e0d40b27a1.tar.xz
  - NB: link is taken from here: 
system-image.ubports.com/ubports-touch/15.04/rc/deb/index.json look for 15.04, deb, rc, v3
  - `tar xf device-1b9*a1.tar.xz`
  - you need the `partitions/boot.img` and the `system/var/lib/lxc/android/system.img` for the following steps
- replace the /var/lib/lxc/android/system.img file on the device:
  - boot into recovery mode 
  - mount the system.img `cd data; mkdir mnt; mount system.img mnt ; `
  - `cd mnt/var/lib/lxc/android`
  - replace the system.img here with the one you just downloaded from the system-image server
- replace the boot.img
  - boot into fastboot mode 
  - `fastboot flash boot boot.img` 
- reboot into UT
=> observe that basically everything works including sound

**Comment by Flohack74** [comment]
@mariogrip what do you think?

**Comment by zubozrout** [comment]
Hi,
well, sound certainly worked back when I was still having N7 deb - but back in the days it was a bit tricky to get this working: https://forums.ubports.com/topic/44/installing-ubuntu-on-nexus-7-2013-deb-lte :)

I've originally used an image from here: http://system-image.tasemnice.eu/ubuntu-touch/devel-proposed/ubuntu/deb/ but it died early in 2016.

Now I feel sad I don't have N7 any more - it was a great device and it was a blast running Ubuntu Touch on it.

**Comment by doniks** [comment]
@zubozrout sound is not an issue. it only doesn't work with the flo image on deb device. sound DOES work in deb device when using the deb image

this issue is just about adding deb to the json files on the server to allow normal installation

**Comment by UniversalSuperBox** [comment]
I have flo, not deb, but it works fine using the old Android compatibility and xenial. I've asked @mariogrip to re-enable building of the flo image.

**Comment by zubozrout** [comment]
Great, I've missed that but both Flo and Deb have appeared here: http://system-image.ubports.com/ubports-touch/16.04/devel/ on Monday :).

**Comment by Flohack74** [comment]
Can you test it and then we can close this ? :)

**Comment by NeoTheThird** [comment]
@Flohack74 
>Can you test it and then we can close this ? :)

It's not done by putting it on the system image server, we also need to add it to the installer. `flo` is already confirmed working, for `deb` i still need a recovery. Or can we use the same @UniversalSuperBox?

**Comment by doniks** [comment]
@NeoTheThird , maybe I'm not understanding your question right, or maybe you're just asking for independent confirmation from UniversalSuperBox, but my workaround is to tell the installer to install flo (which uses the flo recovery) to install onto my deb device. 

FWIW, I've now also checked the `partitions/recovery.img` which is included in the deb device file `/pool/device-1b9d2f7256c3d6c98ba93b66ad48d9ce75833978cd65364df45693e0d40b27a1.tar.xz`, when I diff it against http://cdimage.ubports.com/devices/recovery-flo.img which the installer is using, it tells me it is different, when I boot either of these two recoveries the kernel introduces itself as:

```
uname -a
Linux localhost 3.4.0-5-flo #1-Ubuntu SMP PREEMPT Thu Sep 3 14:21:37 UTC 2015 armv7l GNU/Linux
```

so, in summary, images are different, kernels are the same, both recovery images boot and I can adb shell into them and nothing seems broken and the recovery-flo.img allows the installer to complete successfully

**Comment by NeoTheThird** [comment]
@doniks Thank you, yes, that was my question. So I will go ahead and add deb to the installer. Does GSM work for you @doniks?

**Comment by doniks** [comment]
I don't have a sim card to test. It used to work once upon a time in good ol canonical days when I still had a sim card for it

**Comment by NeoTheThird** [comment]
Okay, you should be able to use the installer for it now. I would appreciate if someone could test it @doniks

**Comment by doniks** [comment]
Well, I get a bit further, but it's not quite right. I think the installer is mistaking it for a flo. 

I start from TWRP recovery
installer is sometimes able to detect that it is a deb, otherwise I select it manually
install 16.04/devel
boots to login screen, I can log in, connect wifi, open apps -> OK 

However, I cannot hear any sound from youtube, nor change the volume form the indicator. The indicator is literally called "indicator-sound" with a cogwheel icon and no controls.

I'm sure this means it installed the android container from flo. When I go back to recovery and replace the /var/libl/lxc/somethingsomethign/system.img with the one for deb (from the device-1b9d2f7256c3d6c98ba93b66ad48d9ce75833978cd65364df45693e0d40b27a1 ) and reboot then I get sound

so I think it somehow mistakes the device for a flo. 

To be clear. flo and deb are close enough so we can use the flo recovery.img for installing, but seems there is a difference in the android container (not sure about boot.img)

**Comment by NeoTheThird** [comment]
Okay, thanks for testing

**Comment by Stefano0101** [comment]
Not working just like before, same output in the Recovery. Just tried the latest installer. 15.04 stable.

**Comment by doniks** [comment]
@Stefano0101 
> same

what exactly is "same"? 

Can you install when you manually select the device?

**Comment by Stefano0101** [comment]
Nope, same issue as closed bug No. 300
It install the recovery, dovnload and push the files, reboot to recovery and fails.

**Comment by doniks** [comment]
ah I see. I don't think #300 and #643 are really duplicates though. 

643 is about *enabling* the installation of the images that are already *existing* on the server onto the deb. The starting point is that those images, including the old kernel image, which I think is still from canonical times are working perfectly fine. Originally *serving* these files from the server and *installing* them in the installer was missing. Now they are served just fine, but somehow the installer ends up installing flo instead of deb.

300 is about the existing flo images not *booting* successfully on the reported flo devices. My best guess is that it is indeed down to the hardware revision as indicated in 300. 

separate problems

**Comment by Stefano0101** [comment]
Exactly what I think. 2 different issues.

**Comment by NeoTheThird** [comment]
I put a different recovery on the server, could you try the installer again?

**Comment by Stefano0101** [comment]
I'll try again.

**Comment by Stefano0101** [comment]
Yeah, I just tested the new installer. NOPE. did not work, even when installer says success! This hardware revision of Nexus 7 is problematic. It again trows errors once the files are pushed to recovery and rebooted to recovery to flash it. I would suggest, you just close this as Nexus 7 is an old device and probably it's just wasting of time to try to fix this. We want to progress on 16.04 and go forward with a ,,NEW DEVICES"  
I don't mind not to have UT on Nexus 7.

**Comment by doniks** [comment]
Tried again now, but it doesn't seem to like that recovery image: 

Fastboot: Unknown error:  sending 'recovery' (12592 KB)...OKAY [  0.426s]writing 'recovery'...FAILED (remote: image is not a boot image)finished. total time: 0.430s

Automatically generated error report 
Version: 0.1.13-beta 
Host OS: linux 
Host Dist: Ubuntu Linux 18.04
Package: Snap 
Host Arch: x64 
Node: v8.2.1 

Error log: http://paste.ubuntu.com/p/8r74kn93mz/

**Comment by doniks** [comment]
Also, I observe now that from ut settings -> updates it seems to forever (I waited mintues) look for an update but never find one ... I think there is also something stuck there.

**Comment by Flohack74** [comment]
@doniks the problem with the updater is not flo/deb specific, I got this sometimes across all devices...

**Comment by doniks** [comment]
@Flohack74 I'm wondering whether you are speaking of the same issue as me. I do not **ever** get a system update. 

multiple weeks. multiple reboots. multiple times  hacking a newer version in with (minimally modified) rootstock script. I'm taking about perfect reproducability. It fails 100% of the time. Is this what you mean? Can you point me to the issue nr. I had a look but couldn't find it.

**Comment by papagunit** [comment]
Came here from #300. Did the process on an Ubuntu 18 machine with a 7 2013 wifi, no issues detecting the device, recovery mode, flashing files, etc. Got to the ubuntu spash screen, and it said "This phone needs restoring from a PC or service center". Eventually discovered that I need to flash it back to kitkat (it was on 6), and then go through the installation process as normal. The first 5 steps in [this git](https://github.com/hawkb/UTCWM_N7_patch) helped me out, while using [this image](http://androidhost.org/dmqht).

Just as an fyi for anyone else having issues. Appears to be few other nexus 7 users out there.

**Comment by NeoTheThird** [comment]
Ok, sorry guys. It seems like deb is actually a little more different than we anticipated. We're going to put a pin in this for now and maybe come back to it later.

**Comment by doniks** [comment]
@NeoTheThird ok, so I'll continue to use my hackish install method. Let me know if there is anything I can do to help.

**Comment by doniks** [comment]
FWIW system-image-cli is not happy: 

```
$ system-image-cli 
b'00ffc806\n'
usage: system-image-cli [-h] [--version] [-C DIRECTORY] [-b BUILD]
                        [-c CHANNEL] [-d DEVICE] [-f FILTER] [-m MAXIMAGE]
                        [-g] [-i] [-n] [-v] [--progress PROGRESS]
                        [-p PERCENTAGE] [--list-channels] [--factory-reset]
                        [--production-reset] [--switch CHANNEL]
                        [--show-settings] [--set KEY=VAL] [--get KEY]
                        [--del KEY] [--override-gsm]
system-image-cli: error: 
Configuration directory not found: /etc/system-image/config.d
```

**Comment by S-Integral** [comment]
@NeoTheNerd, @doniks, @Stefano0101 , @Flohack74 

deb is of particular interest to me.  I do not currently own a N7 WiFi or LTE yet, see you need assistance.  In general, tablets are of greater interest to me than phones.  There are exceptions (LG bootloaders are locked except those shown here: https://developer.lge.com/resource/mobile/RetrieveBootloader.dev).  I have not seen any interest in the N9 either.  Can we discuss here or elsewhere?

S-Integral (Rick)

**Comment by UniversalSuperBox** [comment]
@S-Integral, the forum is the best place for discussions of that type, but in general:

Yes, we need people to maintain the Nexus 7 2013s if good support for them is going to continue.

The Nexus 9 will likely never be supported at this point, it is a 4 year old tablet with a Tegra chipset. Tegra chipsets are notoriously hard to do *anything* with (unless you're Nintendo, apparently). This is why the 2012 Nexus 7 and Nexus 10 fell out of support.

If you'd like to discuss futher, why not open a post in the OS or Porting sections of [our forum](https://forums.ubports.com)?

**Comment by dobey** [comment]
Nexus 10 is Exynos, not Tegra, I think. It was never really unsupported, but it was was basically only ever 2 people had the device to test with, so it wasn't well tested when Canonical was building images. I guess nobody in UBports has one, though.

**Comment by ardalon** [comment]
hi @doniks I'm trying your work around to get it working on my nexus 7 2013 , as the only image i'm availabe to install via the installer is the (flo) version. could you pls explain this step "flash the boot.img from the device-1b.....a1.tar.xz" ?  do you mean write over the boot.img on the tablet with the one in the (deb) image's archive?

**Comment by doniks** [comment]
@ardalon I've added some more details to the first post. let me know whether this is clear

**Comment by doniks** [comment]
New attempt:
1. follow the steps here https://github.com/ubports/docs.ubports.com/pull/196 
1. add `.repo/local_manifests/asus_deb.xml` https://gist.github.com/doniks/e9222f93f3ccab2af70ab45c174ff034 
1. `add_lunch_combo cm_deb-userdebug`
1. `lunch cm_deb-userdebug`
1. compress the kernel more to avoid size error with fastboot flash recovery https://gist.github.com/doniks/cb83c77a423b06ab5e94829f8d2c0471
1. `mka`
1. `fastboot flash boot boot.img` and `fastboot flash recovery recovery.img`
1. `ubuntu-device-flash touch --device=deb --channel=ubports-touch/16.04/rc`
